osu.Game/Screens/SelectV2/BeatmapSetPanel.cs

```diff
@@ -3,63 +3,283 @@
 
 using System;
 using System.Diagnostics;
+using System.Linq;
+using System.Threading;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
 using osu.Framework.Extensions.Color4Extensions;
 using osu.Framework.Graphics;
+using osu.Framework.Graphics.Colour;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Effects;
 using osu.Framework.Graphics.Pooling;
 using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.Sprites;
 using osu.Framework.Input.Events;
+using osu.Framework.Localisation;
 using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Drawables;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Containers;
 using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Overlays;
+using osu.Game.Resources.Localisation.Web;
 using osuTK;
 using osuTK.Graphics;
 
 namespace osu.Game.Screens.SelectV2
 {
     public partial class BeatmapSetPanel : PoolableDrawable, ICarouselPanel
     {
-        public const float HEIGHT = CarouselItem.DEFAULT_HEIGHT * 2;
+        public const float HEIGHT = CarouselItem.DEFAULT_HEIGHT * 1.6f;
+
+        private const float arrow_container_width = 20;
+        private const float difficulty_icon_container_width = 30;
+        private const float corner_radius = 10;
+
+        private const float duration = 500;
 
         [Resolved]
         private BeatmapCarousel carousel { get; set; } = null!;
 
-        private OsuSpriteText text = null!;
+        [Resolved]
+        private OverlayColourProvider colourProvider { get; set; } = null!;
+
+        [Resolved]
+        private BeatmapManager beatmaps { get; set; } = null!;
+
+        [Resolved]
+        private OsuColour colours { get; set; } = null!;
+
+        [Resolved]
+        private BeatmapDifficultyCache difficultyCache { get; set; } = null!;
+
+        private CancellationTokenSource? singleDiffStarsToken;
+        private IBindable<StarDifficulty?>? singleDiffStarsBindable;
+
+        private Container panel = null!;
```

Comment:
Kinda apprehensive at the number of moving pieces in this one class (number of fields and number of sets and update methods). Not sure it's an issue. Maybe better if this gets split out into a case which doesn't also attempt to handle single difficulties. Not sure yet.

Reply:
I was considering doing that but was slightly hesitant about the amount of codelines that will be copied to have the same layout, but it was only yesterday that I realise having a separate component was your plan all along. I'll split it out.